### ISSUE TITLE
fix(types): 数値 cast と expect の防御的修正と型整合化

### ISSUE THREAD

**Issue by thkt** [thkt/rurico#99]
## 背景

`Config::validate()` が境界値を runtime で検証しているが、validated bound を type system 経由で carry できておらず、各箇所で `as` cast や `expect` が「呼び出し前に validate 済」を暗黙に前提する設計になっている。一方で fixture load や `get_local_attention_mask` など、validate を経由しない経路にも同種の cast が散在している。

具体的影響：

- corrupt fixture file で OOM trigger 可能（observable harm 候補）
- 32-bit target で `Config::validate` が silently wrong bound を計算
- 負値 i32 が usize に wrap して panic
- regression 時に production hot path で panic

### fixture load の `as usize` cast

```rust
// src/embed/fixtures.rs:114
let num_docs = read_u32(r)? as usize;
let mut docs = Vec::with_capacity(num_docs);  // u32::MAX 入力で OOM 可能
// 117: let num_chunks = read_u32(r)? as usize;
// 120: let dim = read_u32(r)? as usize;
// 121: let mut bytes = vec![0u8; dim * 4];   // dim * 4 が usize で overflow する 32-bit risk
```

### 負値 i32 → usize wrap

```rust
// src/modernbert/model.rs:436-438
fn get_local_attention_mask(seq_len: i32, half_window: i32) -> Result<Array, Exception> {
    let n = seq_len as usize;        // seq_len < 0 で巨大 usize → vec! で OOM panic
    let hw = half_window as usize;
```

呼び出し側は `Config::validate()` で gate しているが、関数 signature 上は任意の i32 を受け入れる。

### 32-bit ターゲットでの overflow

```rust
// src/modernbert/config.rs:80
let max_local = (i32::MAX as usize).saturating_mul(2).saturating_add(1);
// 32-bit target で usize は 32 bit、saturating_mul(2) は usize::MAX 飽和し意図と乖離
```

### production path の `expect`

```rust
// src/embed/mlx.rs:95
let tokens = tokens_iter.next()
    .expect(\"chunks_per_doc total must match all_chunk_tokens length\");

// src/embed/mlx.rs:300
.map(|slot| slot.expect(\"every chunk slot must be filled by bucket forward\"))

// src/modernbert/model.rs:223-225 ほか多数
let _ = i32::try_from(self.max_seq_len).expect(\"bounded by model config\");
```

invariant が成り立たない distribution bug / regression で production panic。

### 精度損失のある `as f32`

```rust
// src/embed/metrics.rs:137
padded as f32 / real as f32        // usize > 2^24 (~16M) で precision loss
```

production batch で MAX_SEQ_LEN=8192 × TOKEN_BUDGET=256000 累積で接近する範囲。

## アウトカム影響

| 項目                                       | 影響                                                       |
| ------------------------------------------ | ---------------------------------------------------------- |
| corrupt fixture OOM (TYP-004 / RES-014)    | observable, fixture load は `pub fn`                       |
| 32-bit overflow (TYP-021)                  | rurico は 64-bit Mac 主戦場だが target 制限なし           |
| 負値 i32 wrap (TYP-009)                    | 現状 callers は gate、defensive 不在                      |
| production expect panic (TYP-011/012/014)  | regression 時に prod 全停止                                |
| metric precision (TYP-007)                 | padding_ratio の精度に影響、SLA 計測信頼性                |

### 1. fixture load の `as usize` を `try_from` に置換

```rust
// src/embed/fixtures.rs:114-121
let num_docs = usize::try_from(read_u32(r)?)
    .map_err(|_| io::Error::new(io::ErrorKind::InvalidData, \"num_docs overflow\"))?;
// 117 / 120 同様
let bytes_len = dim.checked_mul(4)
    .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidData, \"dim * 4 overflow\"))?;
let mut bytes = vec![0u8; bytes_len];
```

### 2. `get_local_attention_mask` に precondition assert / try_from

```rust
fn get_local_attention_mask(seq_len: i32, half_window: i32) -> Result<Array, Exception> {
    let n = usize::try_from(seq_len).map_err(|_| /* Exception */)?;
    let hw = usize::try_from(half_window).map_err(|_| /* Exception */)?;
    ...
}
```

または signature を `u32` / `usize` に変更（より侵襲的）。

### 3. 32-bit overflow の修正

```rust
// src/modernbert/config.rs:80
// usize 経由を避け、u64 で計算
let max_local: u64 = (i32::MAX as u64) * 2 + 1;
if (self.local_attention as u64) > max_local { ... }
```

または `#[cfg(target_pointer_width = \"64\")]` で 64-bit only と明示し、それ以外でビルドエラー。

### 4. `expect` を Err return / `unreachable!` に置換

- `src/embed/mlx.rs:95` の `expect` → `EmbedError::Inference(\"chunk count mismatch\")` 返却
- `src/embed/mlx.rs:300` 同上
- `src/modernbert/model.rs:223-225` の `try_from(...).expect(\"bounded by config\")` → `Config::validate()` を経由した呼び出し前提を docstring で明示しつつ、defense-in-depth として Err return

### 5. 精度損失 cast の annotation 統一

```rust
// src/embed/metrics.rs:137
#[allow(clippy::cast_precision_loss)]  // padding_ratio: usize→f32 lossy >2^24, monitoring tolerable
let ratio = padded as f32 / real as f32;
```

または f64 計算に変更：

```rust
let ratio = (padded as f64 / real as f64) as f32;
```

`retrieval.rs:297` で既に `#[allow(clippy::cast_precision_loss)]` 採用済、同パターンに揃える。

## 受け入れ条件

- `src/embed/fixtures.rs:114,117,120` の `as usize` が `usize::try_from` に置換
- `src/embed/fixtures.rs:121` の `dim * 4` が `checked_mul`
- `src/modernbert/model.rs:437-438` の負値 i32 wrap 防御
- `src/modernbert/config.rs:80` の 32-bit overflow 解消
- `src/embed/mlx.rs:95,300` および `src/modernbert/model.rs:223-225` の `expect` を Err return / 適切な辞書化
- `src/embed/metrics.rs:137` の精度損失 cast に annotation または f64 計算
- corrupt fixture / 負値 input の単体テスト追加
- `cargo test --workspace`, `cargo clippy --workspace --all-targets --all-features -- -D warnings`, `cargo fmt --check` 全 pass

## 解決対象（Wave 1 findings, /audit 2026-05-01-110240）

- TYP-004 (`src/embed/fixtures.rs:114` u32→usize wrap risk)
- TYP-007 (`src/embed/metrics.rs:137` usize→f32 precision loss)
- TYP-009 (`src/modernbert/model.rs:437-438` 負値 i32 wrap)
- TYP-011 (`src/embed/mlx.rs:95` production expect)
- TYP-012 (`src/embed/mlx.rs:300` slot expect)
- TYP-014 (`src/modernbert/model.rs:223-225` validate-bound expect)
- TYP-021 (`src/modernbert/config.rs:80` 32-bit overflow)
- RES-014 (`src/embed/fixtures.rs:120` fixture OOM, TYP-004 と重複)
- EFF-002 (overlap、別 RC で扱う range なら重複指摘可)

## downstream 追従 (amici / yomu / recall / sae)

API 影響なし（内部実装 / docs / CI のみ）。merge 後、4 リポで `Cargo.toml` の `rurico` rev bump のみで取り込み完了。

- [ ] merge 後、amici / yomu / recall / sae 各リポに rev bump 用追従 issue を作成（既存 rurico 追従 issue があれば統合）

## 工数

実装 30min（mechanical）+ テスト 30min。レビュー別。

**Comment by thkt** [comment]
## /audit 2026-05-06 補強 (feat/issue-93-model-lifecycle 上)

新たな audit run で、本 issue に含めるべき findings を追加で検出。RC-004 (architecture/panic-via-mutex-poison) が root cause として 5 findings を unify する。

### 追加対象 findings (このリストに統合推奨)

| ID | file | 観点 |
|---|---|---|
| TYP-002 | `src/reranker.rs:186` | `scores[0]` unchecked index — `score_batch` の shape check に依存、return type で enforce されない |
| TYP-003 | `src/model_io.rs:41` | `assign_bucket` の `.expect("len exceeds MAX_SEQ_LEN")` — 文書化された precondition だが type で強制されない |
| TYP-008 | `src/embed/mlx.rs:73` | `tokens_iter.next().expect(...)` in `build_indexed_chunks` — over-count case で panic、`debug_assert!` は逆方向のみ |
| SIL-004 | `src/reranker/mlx.rs:54` | `unwrap_or(0)` → `assign_bucket(0)` panic path — outer guard あるが inner に guard なし、TYP-003 と関連 |
| RES-010 | `src/embed/mlx.rs:153` | `i32::try_from(bucket_len).expect(...)` — 将来 `TOKEN_BUDGET` increase で panic risk (defensive) |
| RES-012 | `src/artifacts.rs:394` | `usize::try_from(raw_len).expect("bounded by MAX_HEADER_BYTES")` — 16-bit `usize` portability gap (defensive) |

### Root cause: RC-004 architecture/panic-via-mutex-poison

`expect` panic が `Mutex<EmbedderInner>` 内で発生すると、Rust の `Mutex` が永続的に poison され、共有 `Arc<Embedder>` の全後続呼び出しが `lock poisoned` で失敗。任意の logic bug が **process-lifetime DoS** に格上げされる構造的問題。

`expect` を `Result` 伝播に変える本 issue の方針は、この root cause も同時に解消する。具体的に既存リストの TYP-011/012 (`expect` → Err return) を厳密に守れば RC-004 も解決。

### 関連: このPR (feat/issue-93-model-lifecycle) で先行実装済み

- **OPS-005/006** Mutex poison logging — `lock_inner` の `map_err` に `tracing::error!` 追加済み (poison 観測のみ。本 issue の root fix とは別)
- **このPR の audit snapshot**: `~/.claude/workspace/history/audit-2026-05-06-084602.json`

### 工数見積もり調整

既存の 30min は 8 finding 想定。今回の追加 6 件は SIL-004/RES-010/RES-012 が defensive (5min ずつ)、TYP-002/003/008 が production (10min ずつ) で **+45min 程度** 上振れ見込み。
